app/src/App.tsx

```diff
@@ -60,42 +90,105 @@ export const App = () => {
     },
   });
 
-  const onClickCreateButton = () => {
-    result.mutate();
+  const updateTaskMutation = useUpdateTask(
+    {
+      mutation: 
+      {
+        onSuccess: (result) => client.setQueryData(result.id, result.title)
```

Comment:
setQueryDataでやろうとすると結構ややこしくなるので、
```suggestion
      onSuccess: () => client.invalidateQueries(getListTasksQueryKey()),
```
invalidateQueriesでキャッシュをstaleにして、データを最新に更新すると良いと思います。

Reply:
[リファレンス](https://tanstack.com/query/v4/docs/react/reference/QueryClient#queryclientinvalidatequeries)をみるとデフォルトでは、invalidateQueriesはアクティブなクエリ(定期的にリフェっチしているuseQueryなど？)に対してバックエンドで再リクエストして、キャッシュを更新していると言う認識であっていますか？(あまり理解できていないです)
setQueryDataだとサーバ側で値は更新されている前提でサーバにリクエストせずにキャッシュを更新する(onSuccessの中なのでこれで良いかなと思った)ので、パフォーマンス的には良さそうと思ったのですが、具体的にどのようなややこしいことがあるか教えて頂けると助かります！

Reply:
```tsx
  const updateTaskMutation = useUpdateTask({
    mutation: {
      onSuccess: (result) =>
        client.setQueryData(getListTasksQueryKey(), (prevState: any) => {
          console.log(prevState.data.tasks);
          const newTasks = prevState.data.tasks.map((task: Task) =>
            task.id === result.data.task.id ? result.data.task : task
          );
          return {
            ...prevState,
            data: { tasks: newTasks },
          };
        }),
    },
  });
```
setQueryDataだとこんな感じかと思います。処理が複雑になってしまうので、どっちにするかはトレードオフかと思います 。

Reply:
なるほど、クエリキーについてあまり理解できていなかったです。クエリキーにはキャッシュするものまとめて一つで持っておくのですね。タスクの一覧表示するときデータ取得しているuseTaskList()でクエリキーはgetListTasksQueryKey()としているので、そこでそのキーに対するキャッシュが作られている。だから、そのキャッシュの中を更新する時はgetListTasksQueryKey()で行うって感じですね。

Reply:
よく考えると自分のようにキャッシュをタスク一つずつ管理するのは、タスクが増えた際に良くなさそうですよね